### DISCUSSION TITLE
Dataframes with a column of lists - ui.aggrid can display them but ui.table cannot.

### DISCUSSION THREAD

**Discussion by rohitsathish** [zauberzeug/nicegui#2697] [Q&A]
### Question

If the ui.table line is uncommented, it causes the app to crash, and I've attached a screenshot of how the crash screen looks. Both lists and numpy arrays don't seem to work which is a severe restriction. What are the current limitations of ```ui.table``` with regard to data types?

```python
from nicegui import ui
import pandas as pd
import numpy as np
import ast
    
@ui.page('/')
async def main_ui():
    df = pd.DataFrame({'A': ['[1, 2, 3]', '[4, 5]', '[]']})
    df['A'] = df['A'].apply(lambda x: ast.literal_eval(x))
    #df['A'] = df['A'].apply(lambda x: np.array(x))
    ui.aggrid.from_pandas(df)
    #ui.table.from_pandas(df)

ui.run(port=8081, storage_secret='test')
```

Screenshot: ![Screenshot 2024-03-13 185807](https://github.com/zauberzeug/nicegui/assets/46263358/2eb07922-375e-44e1-80a9-0a6455f4618d)

**Comment by falkoschindler** [comment]
Well, it's not good that the app is crashing. I'm not sure what's going on.
But the NumPy example is working for me. 🤔

**Reply by rohitsathish** [reply]
My mistake, I used ```np.array``` incorrectly, ```df['A'] = df['A'].apply(lambda x: np.array(x))``` outputs a string, not an array. 

Here is the minimal example which outputs an ndarray, and here again only ```ui.aggrid``` works. Its a moot point anyway, since the usual way is to convert to a list and then an ndarray rather than this tedious string manipulation. 

```python
from nicegui import ui
import pandas as pd
import numpy as np
import ast
    
@ui.page('/')
async def main_ui():
    df = pd.DataFrame({'A': ['[1, 2, 3]', '[4, 5]', '[]']})
    #df['A'] = df['A'].apply(lambda x: ast.literal_eval(x))
    df['A'] = df['A'].apply(lambda x: np.array([int(i) for i in x.strip('[]').split(',') if i]))
    print(type(df['A'][0]))
    ui.aggrid.from_pandas(df)
    #ui.table.from_pandas(df)

ui.run(port=8081, storage_secret='test')
```

**Reply by falkoschindler** [reply]
I'm not sure why lists are such a problem for `ui.table`. But I suggest to sanitize the dataframe yourself before creating the table. This way you have more control over how the data is displayed.

```py
df = pd.DataFrame({'A': [[1, 2, 3], [4, 5], []]})
df['A'] = df['A'].apply(lambda x: ', '.join(map(str, x)))
ui.table.from_pandas(df)
```

**Reply by rohitsathish** [reply]
Okay, but I would like to work with and display dataframes with ```ui.table``` that have lists in them. Since I'm doing a lot of data manipulation and iteratively working on the dataframe, its a hassle to keep changing datatypes before I use ```ui.table```.

I guess that still needs to be resolved.

**Reply by rohitsathish** [reply]
@falkoschindler is this important enough to open an issue? Seems quite important for those using nicegui with pandas. Thanks.

**Reply by falkoschindler** [reply]
@rohitsathish Yes, I just created issue https://github.com/zauberzeug/nicegui/issues/2744. Maybe someone from the community can help us finding the root cause.

**Reply by petergaultney** [reply]
@rohitsathish this workaround works for me, assuming there's nothing in particular you need to do with the list once it's in the UI. I _think_ Quasar would have been doing the same on the client side:

```py
        def fixup_lists_because_of_2744(row):
            # https://github.com/zauberzeug/nicegui/issues/2744
            return {k: str(v) if isinstance(v, list) else v for k, v in row.items()}

        table = ui.table(
            columns=columns, rows=list(map(fixup_lists_because_of_2744, rows)),
        )
```

You would have to create the columns and rows manually, which will be slightly more work, but it's not terribly painful to construct the columns data, and the rows would probably just be `df.to_dict(orient='records')`

**Reply by falkoschindler** [reply]
Issue #2744 has been resolved in version 2.18.0. ✔️
